### ISSUE TITLE
Organisation customised templates - section display ordering incorrect 

### ISSUE THREAD

**Issue by tomhigginsuom** [DMPRoadmap/roadmap#1371]
**Expected behaviour:**
When customising a template to add a new section it is possible to change the "Order of display". Setting this to "1" should display the customised section before the built-in template sections.

**Actual behaviour:**
For several of the templates ("AHRC Data Management Plan", "Horizon 2020 DMP" and "MRC Template") the first section displayed is "Data Summary" even though the customised template has an "Order of display" set to "1".

For example here is how the "AHRC Data Management Plan" appears with the "Data Summary" section displayed before the "Manchester Data Management Outline" section:

<img width="1135" alt="data-summary-order-of-display" src="https://user-images.githubusercontent.com/11716796/38618003-520fc2ce-3d90-11e8-97a5-674c1c9da1b1.png">

In addition the sections in the "Template details" tab do not always match the other tabs. For example in the "AHRC Data Management Plan" the first section "Data Summary" is not displayed:

<img width="978" alt="data-summary-order-of-display" src="https://user-images.githubusercontent.com/11716796/38618234-ecd9578e-3d90-11e8-8952-fd5ecdc9041f.png">

When a user views a plan the sections are displayed as follows (with the "Data Summary" section first):

<img width="995" alt="user-plan-section-order" src="https://user-images.githubusercontent.com/11716796/38618697-519a4d76-3d92-11e8-83be-a5daf415cb20.png">

Because this problem appears only on some templates it may be an issue with the template data.

**Steps to reproduce:**

Create a new customised template for "AHRC Data Management Plan", "Horizon 2020 DMP" or "MRC Template" and add a new custom section. Set the "Order of display" for the new section to "1". The new custom section is not displayed first.

**Comment by stephaniesimms** [comment]
my understanding is that customization should only permit admins to add sections to the end of a funder template. can you confirm @sjDCC ? i just tried to customize some other funder templates (Moore foundation, DOE) by adding a section 1 and it appears in the second position after the funder section 1 - definitely a bug with ordering (screenshot below).

we can discuss permitting admins to add a section 1. at minimum we should provide some explanation/tooltip about only being able to add a section to the end.

![screen shot 2018-05-01 at 8 10 22 am](https://user-images.githubusercontent.com/12703638/39478412-396f2e76-4d17-11e8-8087-305287a136e6.png)

**Comment by tomhigginsuom** [comment]
We have been able to add a customised section to the start of some templates - here's an example for ESRC:

<img width="1165" alt="screen shot 2018-05-01 at 17 42 47" src="https://user-images.githubusercontent.com/11716796/39482521-447e9f40-4d67-11e8-9393-f339cf2c1977.png">

**Comment by stephaniesimms** [comment]
thanks for the additional info @tomhigginsuom - it looks like we need to enable addition of a customized section at the beginning or end of a funder template then, and provide guidance about this for admins. we also need to make sure the order of sections behaves correctly, which i don't think is a data problem but may require further investigation.

**Comment by jollopre** [comment]
The current code does not prevent adding phases, sections or questions with duplicate number yet, however we are discussing what would be the best to prevent that. In my opinion, when you customise a template, you should be able to add phases/sections/questions but these should be attached at the end of each collection since funder templates when customised are not meant to be modifiable. 

What are your thoughts on this @sjDCC, @vyruss, @briri, @xsrust ?

**Comment by vyruss** [comment]
Adding a section(/phase/question) numbered "1" when one already exists does not break anything, but those two sections will be presented in "database order". Generally we number sections "0" to make them appear before the first one.

**Comment by sjDCC** [comment]
Yes, Manchester had asked to add their customised sections at the start and we found that by numbering them 0 or 1 they could appear at start. I think now with the dropdown it's not possible to number it as 0 so that is perhaps why some are jumping into second place.

I think the default should be adding custom sections at the end, but I'm happy for institutions to put them in first position too. What we don't want is for it to get mixed up in the funder template section sequence.

What is the best solution to enable these two choices @jollopre @vyruss @briri @xsrust?

**Comment by stephaniesimms** [comment]
@sjDCC we can confirm w/devs on next call if it's possible to constrain order to either "0" (before template sections) or sections at the end.  if not, it sounds like the solution is to only permit adding a section/question at the end of funder templates.

**Comment by Bodacious** [comment]
@sjDCC @stephaniesimms I would suggest we update the view template to allow the user to drag and drop the sections.

- Permitted users will be able to click and drag each section
- Sections will automatically reposition themselves once sorted
- The server will update the sections in the background to store their new position
- Sections will be displayed by order of their sort position.

I'd be happy to implement this now if that suits?

**Comment by sjDCC** [comment]
Thanks @Bodacious 

Just to confirm, the org admins shouldn't have permissions to drag/drop/reorder any of the funder template sections. They should just be able to add their own section to the start or end of this.

Most unis only add one section, but technically they could add more. We should constrain it to one section at the start of the template though. If they have more than one, then additional custom sections should be added to the end of the funder template.

If the drag/drop functionality works well then it could be useful to apply to the standard template editing page too to allow admins to more easily update their content.

Would be good to run through this approach at the dev call tomorrow too to show everyone how things work

**Comment by stephaniesimms** [comment]
i can't create new sections. when i click to create, enter section title and click save I get a success message but there is no new section. there doesn't appear to be a drag/drop functionality or i'm not sure how it's supposed to work. i see a couple of new icons (hamburger and draft in screenshot below) and am not sure what they mean. this issue merits a demo/walk through on a call so we can all understand how it's supposed to work.

**Comment by briri** [comment]
I think this is fixed as part of #1737

**Comment by sjDCC** [comment]
Hi @Bodacious I'm getting a couple of errors with this on roadmap-stage. It won't let me create a section (see below) and even if I set the number to higher than the total number of sections it still won't create.
![capture](https://user-images.githubusercontent.com/6169940/43633430-21e3b752-9701-11e8-81ab-142cad3d256e.PNG)

And it seems existing custom sections aren't being copied through. See what's on preview from previous testing.
![capture2](https://user-images.githubusercontent.com/6169940/43633446-2d0d415c-9701-11e8-9cfd-ed08fbbd956e.PNG)

**Comment by stephaniesimms** [comment]
I noticed the same problems as @sjDCC 
- i can't create a new section. I also wonder if it's appropriate to have the 'Order of display' field at all now? we need to remove this or replace it with something that makes it clear users can only choose Order First or Order Last (screenshot 1)
- existing customized sections are visible in the Template Overview but don't appear in the phase sections (already noted above)
- I propose revising the tooltip text for clarity (screenshot 2). It may work better as a standard popover w/this longer message: "Drag arrows to rearrange customized sections. You may place them before or after the main template sections. " 

![screen shot 2018-08-03 at 9 56 17 am](https://user-images.githubusercontent.com/12703638/43655769-d1e05054-9704-11e8-8c8f-b977c54f3f75.png)

![screen shot 2018-08-03 at 9 57 16 am](https://user-images.githubusercontent.com/12703638/43655770-d1fb5a34-9704-11e8-8f34-6d11d8474f31.png)

**Comment by Bodacious** [comment]
@stephaniesimms 

These issues should be fixed now. 
1) The issue of new sections not being added was caused by the changes to the validations (0d4cdeee0f4023bdfe1b49344031e6a80ee0ae3c) (the system was trying to create the section with the `published` attribute set to nil. This has been fixed since the published attribute was removed (9cd9c9e8ea65a20e80e02da0a59dee0eb37c2957)
2) This should be fixed since the validations were added. All users will have to make sure their data has been cleaned as described in [this README](https://github.com/DMPRoadmap/roadmap/tree/development/lib/data_cleanup). **Developers should not run these until we've merged them into master, since we're still finalising the data cleanup rules**.
3) This message was changed in: [this commit](cphttps://github.com/DMPRoadmap/roadmap/blob/development/app/views/org_admin/phases/container.html.erb#L57-L58)

**Comment by sjDCC** [comment]
Hi @Bodacious New message looks great thanks. I'm still able to replicate the other issues though unfortunately. More details below to help troubleshoot.

When I try to create a new section on a customisable template or a standard template belonging to my org I get the following error. Is it searching for a value but not getting one now that field is removed?
![capture](https://user-images.githubusercontent.com/6169940/43861203-07db45c2-9b4e-11e8-93b0-c4c269d7eb8a.PNG)

As noted above, existing custom sections are showing on preview but not in the actual editable version so I can't trial the drag/drop. If you want to see an example of this, look at American Uni customisation of Arctic Data Centre, which appears to have 4 additional sections.
![au-arcticdatacentre](https://user-images.githubusercontent.com/6169940/43861278-4123dfba-9b4e-11e8-828e-83252f96cc26.PNG)
![edit-view](https://user-images.githubusercontent.com/6169940/43861286-44982548-9b4e-11e8-9c86-516f9ec2b0d5.PNG)

I noticed when testing the drag and drop on an institutional template that the cutomisation logic (i.e. only one before, multiple after) is being applied there too. Or at least it's throwing up a warning when I move sections. Institutions and funders should be able to reorder sections however they like in their own templates.
![drag-drop](https://user-images.githubusercontent.com/6169940/43861492-ed5676c6-9b4e-11e8-8c01-ede383ab8c28.PNG)

@stephaniesimms one usability question this raised for me was being able to see that a change has saved. As there is no section number anymore I couldn't validate the new position. Is it worth having this field but greyed out / uneditable?

**Comment by stephaniesimms** [comment]
I experienced all the same issues noted by @sjDCC above, including when trying to create new data for templates that don't have any existing customizations:
- can't create new sections (error message - number can't be blank)
- can only see existing custom sections on the preview, not while editing the phase
- drag and drop should be not be applied to an org's "Own Templates," only "Customizable Templates" - admins should be able to edit these freely

I'm not sure how to address your question about usability @sjDCC since I can't create and reorder sections, but my feeling is that once drag and drop ordering is working as expected users should have sufficient visual confirmation of their changes/reordering ?

**Comment by sjDCC** [comment]
I managed to drag and drop on the institutional template and only got the error message about one section before funder template sections. If there isn't a visual confirmation that sections have been successfully reordered, we should perhaps add one.

**Comment by sjDCC** [comment]
Retested the drag/drop on DMPonline-dev and looks good. Only issue is related to the organisations own templates. They should be able to completely change the order of these so we want the warning about one section before or multiple after to disappear. Thanks @Bodacious

**Comment by sjDCC** [comment]
I can create sections now - thanks! and the drag and drop works well from a usability perspective. The reordering doesn't seem to be retained though and it's not clear whether changes have saved. I can move sections but on page refresh or navigation elsewhere they jump back to creation order. See BCO-DMO NSF OCE: Biological and Chemical Oceanography example customization.
![1](https://user-images.githubusercontent.com/6169940/44215920-9947eb00-a16b-11e8-9978-8f701417ef4a.PNG)
![2](https://user-images.githubusercontent.com/6169940/44215921-9947eb00-a16b-11e8-80b1-1b58b2e9c051.PNG)

This is the only additional thing I uncovered from the test on DMPonline-dev yesterday. This issue I flagged there that institutions should be able to move any section to any position in their own templates still stands. They shouldn't see this error except when customising funder templates.
![capture](https://user-images.githubusercontent.com/6169940/44216288-81249b80-a16c-11e8-9ded-b6efea273de5.PNG)
